src/rpm.rs

```diff
@@ -1253,7 +1389,137 @@ impl RPMBuilder {
         self
     }
 
-    pub fn build(mut self) -> Result<RPMPackage, RPMError> {
+    /// build without a signature
+    ///
+    /// ignores a present key, if any
+    pub fn build(self) -> Result<RPMPackage, RPMError> {
+        let (lead, header_idx_tag, content) = self.prepare_data()?;
+
+        let mut header = Vec::with_capacity(128);
+        header_idx_tag.write(&mut header)?;
+        let header = header;
+
+        let (
+            _header_len,
+            header_digest_sha1,
+            header_and_content_len,
+            header_and_content_digest_md5,
+        ) = Self::derive_hashes(header.as_slice(), content.as_slice())?;
+
+        let digest_header = Header::<IndexSignatureTag>::builder()
+            .add_digest(
+                header_digest_sha1.as_str(),
+                header_and_content_digest_md5.as_slice(),
+            )
+            .build(header_and_content_len as i32);
+
+        let metadata = RPMPackageMetadata {
+            lead,
+            signature: digest_header,
+            header: header_idx_tag,
+        };
+        let pkg = RPMPackage { metadata, content };
+        Ok(pkg)
+    }
+
+    /// build with the given signer type
+    ///
+    /// The signer will be instantiated as needed if a signing key is present.
+    /// GPG signing key must be in DER / PKCS8 format
+    #[cfg(feature = "signing-meta")]
+    pub fn build_and_sign<S>(self, signing_key: &[u8]) -> Result<RPMPackage, RPMError>
```

Comment:
Thinking about this further, I think you where right in the first place on passing a concrete type instead of a generic byte array. I guess it would be better to just provide a method accepting a `Signer`.
This way we take out all ambiguity out of our api without sacrificing a lot of usability.
Additionally, I think we should not overload the build method and instead add another configuration method like this:
```rust
builder.sign_with(pgp::Signer::load_from(raw_asc)?)
```
This is more a taste thing, but I like it better if a builder has a clear distinction between configuration and construction.

Reply:
The intention here was to keep the signing as versatile as possible without introducing much additional API surface.
Having a `build_with_signer` as well as a `build_and_sign` have the advantage that only the associated functions become generics - otherwise quite a few structs becomes generics which is what I wanted to avoid in the first place by introducing those tow.
Semantically you can also only sign at the very end of building the RPM.

Reply:
> I think the api would be nicer if we would just provide one "build_and_sign" method. Would it be possible to implement the TryFrom crate for Signer? (https://doc.rust-lang.org/std/convert/trait.TryFrom.html) This way, you could accept both a customized signer as well as a byte slice and ditch the "build_with_external_signer" method.

Could be done, I am just not sure making this implicit is a gain from an API POV.

Reply:
> Semantically you can also only sign at the very end of building the RPM, so I did not see that as a downside at all.

This is a valid argument  and I agree.

> Could be done, I am just not sure making this implicit is a gain from an API point of view, I'd rather just remove the one accepting &[u8] entirely.

so we leave it at `build_and_sign(signer)` as an api?

Reply:
Let's remove the variant taking a serialized key `&[u8]` and keep the one taking a generic impl of `Signer`.